File: assets/js/analyses/lab-ledger.js
/* assets/js/analyses/lab-ledger.js */
(function () {
  "use strict";
  (window.AnalysesRegistry = window.AnalysesRegistry || {
    _q: [], register: function (s, d) { this._q.push([s, d]); }
  }).register("lab-ledger", {
    prose: {
      intro:
        "<p>Collapse everyone in the three labs into their lab — the unaffiliated periphery " +
        "holds no edges here — and 312 coauthorship edges " +
        "become a nine-number ledger: how much each lab trades with itself and with the " +
        "other two. The question is whether any of that trade is more than what each " +
        "lab&rsquo;s sheer size and star power would produce on their own. The two-block " +
        "reading is already visible in the raw counts: of the 312 edges, not one runs " +
        "directly between EleutherAI and the Vogelstein lab, while the interpretability " +
        "seam between EleutherAI and the Bau lab carries 37 direct edges with a summed " +
        "tie strength of 5.4.</p>",
      how:
        "<p>Raw totals mislead here, because a lab full of prolific hubs racks up " +
        "cross-lab weight by volume alone — the way a high-norm activation dominates a " +
        "raw dot product whether or not it points anywhere interesting. So we normalize: " +
        "each cell of the ledger is divided by the weight you&rsquo;d expect if every " +
        "person kept exactly their total collaboration weight but rewired <em>who</em> they " +
        "collaborate with completely at random.</p>" +
        "<p>That denominator is the degree correction in a degree-corrected stochastic " +
        "blockmodel — the graph analogue of normalizing activation magnitudes before " +
        "comparing directions, cosine similarity instead of raw dot products. What " +
        "survives is pure preference: a cell at 1.0 means hub sizes alone explain the " +
        "traffic, above 1.0 means the labs seek each other out, below 1.0 means a wall. " +
        "On this map the diagonal glows at 2.3&ndash;4.5&times; chance and almost everything " +
        "off it sits far below — except one channel.</p>",
      method:
        "<p>Degree-corrected stochastic blockmodel block matrix &mdash; Karrer &amp; " +
        "Newman, &ldquo;Stochastic blockmodels and community structure in networks&rdquo;, " +
        "<em>Physical Review E</em> 83, 016107 (2011) &mdash; with the block assignment " +
        "fixed to the page&rsquo;s three communities rather than re-fit. Observed " +
        "W<sub>ab</sub> sums edge weights between labs; edges use fractional " +
        "co-authorship counting &mdash; each paper adds 1/n<sub>authors</sub> to each of " +
        "its co-author pairs (suggested by Stella Biderman) &mdash; so one many-author " +
        "paper no longer counts as several units: the EleutherAI&ndash;Bau cell&rsquo;s " +
        "5.4 units of tie strength come from 21 distinct papers. The null is the " +
        "weighted configuration model, E<sub>ab</sub> = d<sub>a</sub>d<sub>b</sub>/2m off " +
        "the diagonal and d<sub>a</sub><sup>2</sup>/4m on it, so the printed ratio " +
        "W<sub>ab</sub>/E<sub>ab</sub> is proportional to the DCSBM mixing-matrix MLE. " +
        "The ledger&rsquo;s thinnest channel, Bau&ndash;Vogelstein, holds three recorded " +
        "papers &mdash; Agents of Chaos, plus NNsight/NDIF counted twice under two title " +
        "variants the databases never merged &mdash; and every one of its ten edges " +
        "touches Alexander R. Loftus, whom this re-clustering itself moved to the " +
        "Vogelstein side. Caveats: the communities were clustered from this same graph, " +
        "so diagonal ratios above 1 are partly by construction &mdash; the off-diagonal " +
        "comparisons are the informative part; and a zero is a statement about what the " +
        "databases recorded, never about the people.</p>"
    },
    render: function (el, data, shared) {
      var d = data.data;
      var labs = d.labs;
      var C = shared.colors;
      var labColor = function (i) { return C.community[labs[i].id % C.community.length]; };

      var maxRatio = d3.max(d.ratio.flat());
      var tint = d3.scaleLinear().domain([0, maxRatio]).range([0, 0.2]);
      var fmtRatio = function (r) {
        if (r === 0) return "0";
        if (r >= 10) return r.toFixed(0);
        if (r >= 1) return r.toFixed(1);
        if (r >= 0.095) return r.toFixed(2);
        return r.toFixed(3);
      };

      var w = el.clientWidth || 720;
      var labelCol = 116;
      var avail = Math.min(w, 760) - labelCol - (w >= 640 ? 240 : 0);
      var cellW = Math.max(72, Math.min(120, Math.floor(avail / 3) - 6));
      var cellH = Math.max(52, Math.round(cellW * 0.6));

      var root = d3.select(el).append("div")
        .style("display", "flex").style("flex-wrap", "wrap")
        .style("gap", "12px 36px").style("align-items", "flex-start");

      // ---- corner anchor: the standard mini-map, colored by lab -----------------
      var right = root.append("div").style("order", "2");
      var mini = shared.minimap(right.node(), function (id) {
        var c = shared.communityOf(id);
        return c >= 0 ? C.community[c % C.community.length] : null;
      });
      right.append("div").attr("class", "m-note").style("max-width", "200px")
        .text("The three labs, same colors as the ledger. Hover a cross-lab cell to " +
          "light up its biggest carrier.");

      // ---- the 3×3 table-graphic: the printed ratios ARE the ink ----------------
      var left = root.append("div").style("order", "1");
      var grid = left.append("div")
        .style("display", "grid")
        .style("grid-template-columns", labelCol + "px repeat(3, " + cellW + "px)")
        .style("gap", "5px");

      grid.append("div"); // empty corner
      labs.forEach(function (lab, j) {
        grid.append("div").attr("class", "m-label")
          .style("color", labColor(j))
          .style("align-self", "end").style("text-align", "center")
          .style("padding-bottom", "2px")
          .text(lab.short);
      });

      var carrierTip = function (i, j) {
        var list = d.carriers[Math.min(i, j) + "" + Math.max(i, j)];
        if (!list || !list.length) return "";
        var top = list[0];
        if (top.share === 1) {
          return "<br><span style=\"color:" + C.muted + "\">every link in this cell " +
            "touches </span><strong>" + shared.esc(shared.labelOf(top.id)) + "</strong>";
        }
        return "<br><span style=\"color:" + C.muted + "\">most tie strength carried by</span> " +
          list.map(function (p) {
            return shared.esc(shared.labelOf(p.id)) + " (" + shared.fmt.num(p.w) + ")";
          }).join(", ");
      };
      var cellTip = function (i, j) {
        var within = i === j;
        var head = within
          ? "within " + shared.esc(labs[i].short)
          : shared.esc(labs[i].short) + " &harr; " + shared.esc(labs[j].short);
        var nPairs = within ? labs[i].n * (labs[i].n - 1) / 2 : labs[i].n * labs[j].n;
        var html = "<strong>" + head + "</strong><br>" +
          "tie strength " + shared.fmt.num(d.observed[i][j]) +
          (d.observed[i][j] > 0
            ? " &mdash; " + d.papers[i][j] + " distinct papers across " +
              d.edges[i][j] + " direct edges"
            : "") +
          "<br>random rewiring expects " + shared.fmt.num(d.expected[i][j]) +
          " &rarr; ratio <strong>" + fmtRatio(d.ratio[i][j]) + "</strong><br>" +
          "<span style=\"color:" + C.muted + "\">tie strength " +
          shared.fmt.sig(d.percap[i][j], 2) +
          " per possible pair (" + shared.fmt.num(nPairs) + " pairs)</span>";
        if (d.observed[i][j] === 0) {
          html += "<br><span style=\"color:" + C.muted + "\">no recorded papers — " +
            "every path between these labs runs through the Bau-lab side of the map</span>";
        } else if (!within) {
          html += carrierTip(i, j);
        }
        return html;
      };
      var topCarrierId = function (i, j) {
        if (i === j || d.observed[i][j] === 0) return null;
        var list = d.carriers[Math.min(i, j) + "" + Math.max(i, j)];
        return list && list.length ? list[0].id : null;
      };

      labs.forEach(function (rowLab, i) {
        var rl = grid.append("div")
          .style("align-self", "center").style("text-align", "right")
          .style("padding-right", "10px").style("line-height", "1.25");
        rl.append("div").attr("class", "m-label")
          .style("color", labColor(i)).text(rowLab.short);
        rl.append("div").attr("class", "m-axis").style("color", C.muted)
          .text(rowLab.n + " people");

        labs.forEach(function (colLab, j) {
          var r = d.ratio[i][j];
          var cell = grid.append("div")
            .style("height", cellH + "px")
            .style("display", "flex")
            .style("align-items", "center").style("justify-content", "center")
            .style("background", function () {
              var ink = d3.color(C.ink);
              return "rgba(" + ink.r + "," + ink.g + "," + ink.b + "," + tint(r).toFixed(3) + ")";
            })
            .style("border", i === j ? "1px solid " + C.hair : "1px solid transparent")
            .style("border-radius", "2px")
            .style("cursor", "default");
          var num = cell.append("div")
            .style("font-size", "18px").style("font-weight", "600")
            .style("color", r === 0 ? C.muted : C.ink)
            .style("font-variant-numeric", "tabular-nums")
            .text(fmtRatio(r));
          if (r !== 0) {
            num.append("span")
              .style("font-size", "12px").style("font-weight", "400")
              .style("color", C.muted).text("×");
          }
          cell.on("mousemove", function (evt) {
            shared.tooltip.show(cellTip(i, j), evt);
            mini.highlight(topCarrierId(i, j));
          }).on("mouseleave", function () {
            shared.tooltip.hide();
            mini.highlight(null);
          });
        });
      });

      left.append("div").attr("class", "m-note")
        .style("max-width", (labelCol + 3 * cellW + 15) + "px")
        .style("margin-top", "10px")
        .html("<strong>1.0</strong> = exactly what hub sizes alone predict. Darker = " +
          "further above chance; diagonal (outlined) = within-lab.");
    }
  });
})();

File: experiments/coauthorship/analyses/lab-ledger.py
# /// script
# requires-python = ">=3.10"
# dependencies = ["numpy", "networkx", "scikit-learn", "graspologic==3.4.4", "setuptools<81"]
# ///
"""lab-ledger — DCSBM block matrix: lab-to-lab trade after degree correction.
Run: cd experiments/coauthorship/analyses && uv run lab-ledger.py"""
import json
import re
from collections import defaultdict
from pathlib import Path

HERE = Path(__file__).resolve().parent
REPO = HERE.parents[2]
GRAPH = json.loads((REPO / "public/assets/data/coauthorship.json").read_text())
DERIVED = HERE / "_derived"  # unused here; kept per contract template
OUT = REPO / "public/assets/data/analyses" / "lab-ledger.json"

SLUG = "lab-ledger"
LABS = [0, 1, 2]
SHORT = {0: "EleutherAI", 1: "Bau lab", 2: "Vogelstein lab"}

assert GRAPH["meta"]["weighting"] == "fractional", "expected fractional edge weights"

community = {n["id"]: n["community"] for n in GRAPH["nodes"]}
label = {n["id"]: n["label"] for n in GRAPH["nodes"]}
assert len(community) == len(GRAPH["nodes"]), "duplicate node ids"
lab_label = {c["id"]: c["label"] for c in GRAPH["communities"]}
assert set(lab_label) == set(LABS), f"unexpected community ids {lab_label}"

n_members = {a: sum(1 for c in community.values() if c == a) for a in LABS}
# sizes include the Feucht editorial override (overrides.json: she is a Bau-lab student)
assert n_members == {0: 22, 1: 56, 2: 29}, f"community sizes shifted: {n_members}"

# ---- observed block matrix W (exclude community -1; diagonal counted once) ----------
# W sums fractional edge weights (each paper adds 1/n_authors per pair).
# edges_n counts integer direct edges per cell; cell_papers tracks distinct titles.
W = [[0.0] * 3 for _ in range(3)]
edges_n = [[0] * 3 for _ in range(3)]
cell_papers = defaultdict(set)  # (a,b) a<=b -> distinct paper titles
carrier_w = defaultdict(lambda: defaultdict(float))  # (a,b) a<b -> person -> weight
n_excluded = 0
for l in GRAPH["links"]:
    cu, cv = community[l["source"]], community[l["target"]]
    assert l["weight"] > 0
    assert l["n_papers"] >= 1
    if cu < 0 or cv < 0:
        n_excluded += 1
        continue
    a, b = sorted((cu, cv))
    W[a][b] += l["weight"]
    edges_n[a][b] += 1
    if a != b:
        W[b][a] += l["weight"]
        edges_n[b][a] += 1
        carrier_w[(a, b)][l["source"]] += l["weight"]
        carrier_w[(a, b)][l["target"]] += l["weight"]
    cell_papers[(a, b)].update(l["papers"])
assert n_excluded == 0, "periphery (-1) nodes hold edges now — update prose caveats"
assert all(W[a][b] == W[b][a] for a in LABS for b in LABS), "W not symmetric"
n_cell_edges = sum(edges_n[a][b] for a in LABS for b in LABS if a <= b)
assert n_cell_edges == len(GRAPH["links"]), "edge count mismatch vs shipped graph"

# ---- degree-corrected expectation: E_ab = d_a d_b / 2m off-diag, d_a^2 / 4m on ------
deg = [2 * W[a][a] + sum(W[a][b] for b in LABS if b != a) for a in LABS]
m = sum(W[a][b] for a in LABS for b in LABS if a <= b)
assert abs(sum(deg) - 2 * m) < 1e-9, "handshake check failed"
E = [
    [deg[a] * deg[b] / (2 * m) if a != b else deg[a] ** 2 / (4 * m) for b in LABS]
    for a in LABS
]
R = [[W[a][b] / E[a][b] for b in LABS] for a in LABS]
percap = [
    [
        W[a][b] / (n_members[a] * n_members[b])
        if a != b
        else W[a][a] / (n_members[a] * (n_members[a] - 1) / 2)
        for b in LABS
    ]
    for a in LABS
]
papers_n = [[len(cell_papers[tuple(sorted((a, b)))]) for b in LABS] for a in LABS]
print(f"[{SLUG}] W={[[round(x, 4) for x in row] for row in W]}")
print(f"[{SLUG}] deg={[round(x, 4) for x in deg]} m={round(m, 4)}")
print(f"[{SLUG}] ratio={[[round(x, 3) for x in row] for row in R]}")
print(f"[{SLUG}] percap={[[round(x, 5) for x in row] for row in percap]}")
print(f"[{SLUG}] direct edges per cell={edges_n}")
print(f"[{SLUG}] distinct papers per cell={papers_n}")

# ---- top carriers of each nonzero off-diagonal cell ---------------------------------
carriers = {}
for (a, b), per_person in sorted(carrier_w.items()):
    total = W[a][b]
    top = sorted(per_person.items(), key=lambda kv: (-kv[1], kv[0]))[:3]
    carriers[f"{a}{b}"] = [
        {"id": pid, "w": round(w, 3), "share": round(w / total, 4)} for pid, w in top
    ]
    for pid, _ in top:
        assert pid in community, f"carrier {pid} not in GRAPH nodes"

# ---- verify the headline facts -------------------------------------------------------
assert W[0][2] == 0.0, "EleutherAI–Vogelstein cell no longer zero — rewrite headline"
assert edges_n[0][2] == 0, "direct EleutherAI–Vogelstein edges appeared — rewrite prose"
fold = R[0][1] / R[1][2]
assert 15.0 <= fold < 40.0, f"seam fold left its regime — rewrite headline: {fold}"
top01 = carriers["01"][0]
assert top01["id"] == "kola ayonrinde", f"unexpected top carrier {top01}"
assert 0.20 <= top01["share"] <= 0.55, f"top-carrier share left its regime: {top01}"
c12 = carriers["12"][0]
assert c12["id"] == "alex loftus" and c12["share"] == 1.0, c12
# evidence-line numbers — the JS intro prose bakes these; keep it in sync with this print
print(f"[{SLUG}] seam evidence: edges={edges_n[0][1]} weight={W[0][1]:.2f} papers={papers_n[0][1]}")

headline = (
    "Strip out what hub sizes alone would produce and one open channel dominates the "
    f"ledger: EleutherAI and the Bau lab trade <strong>{round(fold)}&times;</strong> more "
    f"intensely than the Bau and Vogelstein labs do &mdash; and "
    f"<strong>{label[top01['id']]}</strong> alone accounts for "
    f"{round(100 * top01['share'])}% of it."
)

payload = {
    "slug": SLUG,
    "title": "The lab-to-lab ledger",
    "headline": headline,
    "data": {
        "labs": [
            {
                "id": a,
                "label": lab_label[a],
                "short": SHORT[a],
                "n": n_members[a],
                "degree": round(deg[a], 2),
            }
            for a in LABS
        ],
        "m": round(m, 2),
        "observed": [[round(x, 2) for x in row] for row in W],
        "edges": edges_n,
        "expected": [[round(x, 1) for x in row] for row in E],
        "ratio": [[round(x, 3) for x in row] for row in R],
        "percap": [[round(x, 5) for x in row] for row in percap],
        "papers": papers_n,
        "carriers": carriers,
        "fold": round(fold, 1),
    },
}

blob = json.dumps(payload, separators=(",", ":"))
assert len(blob) < 300_000, f"payload too big: {len(blob)}"
OUT.write_text(blob)
print(f"[{SLUG}] OK {len(blob)/1024:.0f}KB — " + re.sub(r"<[^>]+>", "", headline))
